File: odroid/Share/src/bert/share/model/AbstractRobotModel.java
/**
 * Copyright 2018-2019. Charles Coughlin. All Rights Reserved.
 *                 MIT License.
 */
package bert.share.model;

import java.io.IOException;
import java.lang.System.Logger.Level;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import bert.share.motor.MotorConfiguration;
import bert.share.xml.XMLUtility;

/**
 *  This is the base class for a collection of models that keep basic configuration
 *  information, all reading from the same file. The information retained is specific
 *  to the application.
 */
public abstract class AbstractRobotModel  {
	private static final String CLSS = "AbstractRobotModel";
	protected final Document document;
	protected final Properties properties;
	protected final Map<String,String> controllerTypes;  // Controller types by controller name
	protected final Map<String,String> pipeNames;        // Pipe names by controller name
	protected final Map<String,MotorConfiguration> motors; // Motor configuration by motor name
	private static final System.Logger LOGGER = System.getLogger(CLSS);

	
	public AbstractRobotModel(Path configPath) {
		this.document = analyzePath(configPath);
		this.properties = new Properties();
		this.controllerTypes = new HashMap<>();
		this.motors = new HashMap<>();
		this.pipeNames = new HashMap<>();
	}
    

	/**
	 *  Each application needs to extract the controller(s) of interest from the 
	 *  configuration file. Presumably this method will be called as part of the 
	 *  populate() process. 
	 */
	public abstract void analyzeControllers();
	/**
	 *  Analyze the document. The information retained is dependent on the context
	 *  (client or server). This must be called before the model is accessed.
	 */
	public abstract void populate();
	
	public String getProperty(String key,String defaultValue) {
		return this.properties.getProperty(key,defaultValue);
	}
	
	/**
	 * @return a map of type names for each controller used by this application.
	 */
	public Map<String,String> getControllerTypes() { return this.controllerTypes; }
	/**
	 * @return a map of motor configuration objects by motor name (upper case).
	 */
	public Map<String,MotorConfiguration> getMotors() { return this.motors; }
	/**
	 * @return a map of names of the pipes for each controller used by this application.
	 *         The key list is sufficient to get the controller names.
	 */
	public Map<String,String> getPipeNames() { return this.pipeNames; }
    /**
	 * Expand the supplied path as the configuration XML file.
	 * @return the configuration, an XML document.
	 */
	private Document analyzePath(Path filePath) {
		Document contents = null;
		try {
			byte[] bytes = Files.readAllBytes(filePath);
			if( bytes!=null ) {
				contents = XMLUtility.documentFromBytes(bytes);
			}
		}
		catch( IOException ioe) {
			LOGGER.log(Level.ERROR, String.format("%s.getConfiguration: Failed to read file %s (%s)",
											CLSS,filePath.toAbsolutePath().toString(),ioe.getLocalizedMessage()));
		}
		return contents;
	}
	

    // ================================ Auxiliary Methods  ===============================

	
	/**
	 * Search the model for property elements. The results are saved in the properties member.
	 * Call this if the model has any properties of interest.
	 */
	protected void analyzeProperties() {
		if( this.document!=null ) {
			NodeList elements = document.getElementsByTagName("property");
			int count = elements.getLength();
			int index = 0;
			while(index<count) {
				Node propertyNode = elements.item(index);
				String key = XMLUtility.attributeValue(propertyNode, "name");
				if( key==null ) {
					LOGGER.log(Level.WARNING,String.format("%s.analyzeProperties: Missing name attribute in property",CLSS));
				}
				else {
					String value = propertyNode.getTextContent();
					if( value!=null && !value.isBlank() ) {
						properties.put(key.toLowerCase(), value);
					}
				}
				index++;
			}
		}
	}
	
	/**
	 * Search the model for controller elements with joint sub-elements. The results form a list
	 * of MotorConfiguration objects.
	 */
	protected void analyzeMotors() {
		if( this.document!=null ) {
			NodeList controllers = document.getElementsByTagName("controller");
			int count = controllers.getLength();
			int index = 0;
			while(index<count) {
				Node controllerNode = controllers.item(index);
				String type = XMLUtility.attributeValue(controllerNode, "type");
				if( type!=null && type.equalsIgnoreCase("JOINT")) {
					String controller = XMLUtility.attributeValue(controllerNode, "name");
					if( controller!=null ) {
						Node node = controllerNode.getFirstChild();
						while( node!=null ) {
							if( node.getNodeType()==Node.ELEMENT_NODE ) {
								Element joint = (Element)node;
								if( joint.getTagName().equals("joint") ) {
									MotorConfiguration motor = new MotorConfiguration();
									motor.setController(controller);
									String value = XMLUtility.attributeValue(joint, "name");
									if( value!=null) motor.setName(value);
									value = XMLUtility.attributeValue(joint, "type");
									if( value!=null) motor.setType(value);
									value = XMLUtility.attributeValue(joint, "id");
									if( value!=null) motor.setId(Integer.parseInt(value));
									value = XMLUtility.attributeValue(joint, "offset");
									if( value!=null) motor.setOffset(Double.parseDouble(value));
									value = XMLUtility.attributeValue(joint, "min");
									if( value!=null) motor.setMinAngle(Double.parseDouble(value));
									value = XMLUtility.attributeValue(joint, "max");
									if( value!=null) motor.setMaxAngle(Double.parseDouble(value));
									value = XMLUtility.attributeValue(joint, "speed");
									if( value!=null) motor.setSpeed(Double.parseDouble(value));
									value = XMLUtility.attributeValue(joint, "torque");
									if( value!=null) motor.setTorque(Double.parseDouble(value));
									value = XMLUtility.attributeValue(joint, "orientation");
									if( value!=null) {
										motor.setIsDirect(value.equalsIgnoreCase("direct"));
									}
								}
							}
							node = node.getNextSibling();
						}
					}
					else {
						LOGGER.log(Level.WARNING,String.format("%s.analyzeProperties: Missing name attribute in property",CLSS));
					}
				}
				
				index++;
			}
		}
	}
}

File: Bert/common/src/main/kotlin/chuckcoughlin/bert/common/model/AbstractRobotModel.kt
/**
 * Copyright 2022-2023. Charles Coughlin. All Rights Reserved.
 * MIT License.
 */
package chuckcoughlin.bert.common.model

import chuckcoughlin.bert.common.util.XMLUtility
import org.w3c.dom.Document
import org.w3c.dom.Element
import org.w3c.dom.Node
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.*
import java.util.logging.Logger

/**
 * This is the base class for a collection of models that keep basic configuration
 * information, all reading from the same file. The information retained is specific
 * to the application.
 */
abstract open class AbstractRobotModel(configPath: Path) {
    protected val document: Document?
    protected val properties: Properties
    protected val handlerTypes : MutableMap<String, String>   // Map of type names for each message handler used by this application.
    val sockets : MutableMap<String, Int>           // Socket port by handler name. The list keys are sufficient to get the controller names.
    protected val motors : MutableMap<Joint, MotorConfiguration> // Motor configuration by joint

    /**
     * Each application needs to extract the definitiona for controller(s) of interest
     * from the configuration file. Presumably this method will be called as part of the
     * populate() process.
     */
    abstract fun analyzeControllers()

    /**
     * Analyze the document. The information retained is dependent on the context
     * (client or server). This must be called before the model is accessed.
     */
    abstract fun populate()

    fun getProperty(key: String?, defaultValue: String?): String {
        return properties.getProperty(key, defaultValue)
    }

    /**
     * Expand the supplied path as the configuration XML file.
     * @return the configuration, an XML document.
     */
     protected fun analyzePath(filePath: Path): Document? {
        var contents: Document? = null
        try {
            val bytes = Files.readAllBytes(filePath)
            if (bytes != null) {
                contents = XMLUtility.documentFromBytes(bytes)
            }
        }
        catch (ioe: IOException) {
            LOGGER.severe(String.format("%s.getConfiguration: Failed to read file %s (%s)",
                    CLSS, filePath.toAbsolutePath().toString(), ioe.localizedMessage) )
        }
        return contents
    }
    // ================================ Auxiliary Methods  ===============================
    /**
     * Search the model for property elements. The results are saved in the properties member.
     * Call this if the model has any properties of interest.
     */
      open fun analyzeProperties() {
        if (document != null) {
            val elements = document.getElementsByTagName("property")
            val count = elements.length
            var index = 0
            while (index < count) {
                val propertyNode = elements.item(index)
                val key = XMLUtility.attributeValue(propertyNode, "name")
                val value = propertyNode.textContent
                if (value != null && !value.isEmpty()) {
                    properties[key.lowercase(Locale.getDefault())] = value
                }
                index++
            }
        }
    }

    /**
     * Search the XML description for controller elements with joint sub-elements. The results form a list
     * of MotorConfiguration objects.
     */
    protected fun analyzeMotors() {
        if (document != null) {
            val controllers = document.getElementsByTagName("controller")
            val count = controllers.length
            var index = 0
            while (index < count) {
                val controllerNode = controllers.item(index)
                val type = XMLUtility.attributeValue(controllerNode, "type")
                if( type.equals("SERIAL", ignoreCase = true )) {
                    val controller = XMLUtility.attributeValue(controllerNode, "name")
                        var node = controllerNode.firstChild
                        while (node != null) {
                            if (node.nodeType == Node.ELEMENT_NODE) {
                                val joint = node as Element
                                if (joint.tagName == "joint") {
                                    val j = Joint.valueOf(XMLUtility.attributeValue(joint, "name"))
                                    val id = XMLUtility.attributeValue(joint, "id").toInt()
                                    val typ = DynamixelType.valueOf(type.uppercase(Locale.getDefault()))
                                    var value = XMLUtility.attributeValue(joint, "orientation")
                                    val isDirect = value.equals("direct", ignoreCase = true)

                                    val motor = MotorConfiguration(j,typ,id,controller,isDirect)
                                    motor.offset = XMLUtility.attributeValue(joint, "offset").toDouble()
                                    // The following attributes are optional
                                    value = XMLUtility.attributeValue(joint, "min")
                                    if(!value.isEmpty() ) motor.minAngle = value.toDouble()
                                    value = XMLUtility.attributeValue(joint, "max")
                                    if( !value.isEmpty()) motor.maxAngle = value.toDouble()
                                    value = XMLUtility.attributeValue(joint, "speed")
                                    if( !value.isEmpty()) motor.maxSpeed = value.toDouble()
                                    value = XMLUtility.attributeValue(joint, "torque")
                                    if( !value.isEmpty()) motor.maxTorque = value.toDouble()
                                    value = XMLUtility.attributeValue(joint, "limb")
                                    if( !value.isEmpty()) {
                                        try {
                                            val limb = Limb.valueOf(value.uppercase(Locale.getDefault()))
                                            motor.limb = limb
                                        }
                                        catch (iae: IllegalArgumentException) {
                                            LOGGER.warning( String.format(
                                                    "%s.analyzeMotors: %s has unknown limb %s",
                                                    CLSS,motor.joint.name, value ))
                                        }
                                    }
                                    motors[motor.joint] = motor
                                    LOGGER.fine(String.format("%s.analyzeMotors: Found %s", CLSS, motor.joint.name))
                                }
                            }
                            node = node.nextSibling
                        }
                }
                index++
            }
        }
    }

    private val CLSS = "AbstractRobotModel"
    private val LOGGER = Logger.getLogger(CLSS)

    init {
        document = analyzePath(configPath)
        properties = Properties()
        handlerTypes              = mutableMapOf<String,String>()
        motors                    = mutableMapOf<Joint, MotorConfiguration>()
        sockets                   = mutableMapOf<String,Int>()
    }
}